THU-10 Proxy fetch content tool for user privacy

robbyrob42 opened the pull request:
draft pr

robbyrob42 committed:
Add user story documentation structure

Added .claude/user-stories/ directory with THU-10 user story for implementing privacy-protected fetch content tool via Exa proxy.

robbyrob42 committed:
Implement privacy-protected fetch content tool using Exa proxy

Replace direct HTTP fetching with Exa's /contents endpoint to protect user privacy. This prevents ad networks from tracking user IP addresses during web content fetching.

Key changes:
- Add ExaContentFetcher class using Exa's privacy-protected API
- Update /fetch-content endpoint to use Exa proxy instead of direct requests
- Add comprehensive test suite verifying no user data is forwarded
- Maintain same frontend API for backward compatibility

Privacy benefits:
- User IP addresses no longer exposed to target websites
- No identifying headers forwarded to external sites
- All requests proxy through Exa's infrastructure
- Bot blockers and rate limiting handled automatically

Fixes THU-10

robbyrob42 committed:
Update THU-10 user story with implementation documentation

Add comprehensive implementation summary including changes made, privacy verification results, and technical benefits achieved.

cjroth commented:
Just a heads up - we'll want to split the markdown docs stuff into a separate PR so we can review and merge it separately

robbyrob42 commented:
> Just a heads up - we'll want to split the markdown docs stuff into a separate PR so we can review and merge it separately

ah ok, i'll separate it.

robbyrob42 commented:
THU-10 Ready for Review, Thank you!

robbyrob42 committed:
Add WebContentFetcher fallback for improved UX

- Update fetch-content endpoint to use WebContentFetcher when Exa is unavailable
- Maintains privacy-first approach by trying Exa first
- Falls back to direct fetching if Exa not configured or fails
- Add comprehensive tests for both fallback scenarios
- Update endpoint documentation to reflect fallback behavior

This ensures the app always works while prioritizing privacy when possible.

robbyrob42 commented:
Added WebContentFetcher fallback to fetch-content endpoint. Uses Exa proxy when available, falls back to direct fetching when unavailable or failing. Tests added for both scenarios.

cjroth commented on backend/pro/exa_content_fetcher.py:
```diff
@@ -0,0 +1,128 @@
+"""
```
I think we should just have one class for exa - not separate fro content and search.

robbyrob42 replied:
Reasonable, I almost went that direction but wanted to get it working. I'm happy to refactor

cjroth commented on src/ai/fetch.ts:
```diff
@@ -203,13 +203,14 @@ export const aiFetchStreamingResponse = async ({
           console.groupEnd()
```
We need to get this PR to only show the changes relevant to your branch - this is hard to review

robbyrob42 replied:
yeah. agreed, apologies, i'll get my process tightened up.

cjroth commented:
Hey - I think this will be ready for review soon, but it's not there yet:
 - PR description should link to Linear task or be more descriptive
 - Checks should be passing
 - Merge conflicts should be resolved
 - There should only be changes that are part of this branch when I click on the "Files Changed" tab
 - Reviewer should be tagged in the top right
 - Should be marked as "Ready for review" so that it's not in draft state

robbyrob42 committed:
replacing lost ignores for local dev files

Signed-off-by: Rob Jacques <2531474+robbyrob42@users.noreply.github.com>

robbyrob42 committed:
Refactor: Consolidate Exa functionality into unified ExaClient

Consolidate ExaSearcher and ExaContentFetcher into single ExaClient class
for better maintainability and reduced code duplication.

Key Changes:
- Merge ExaSearcher and ExaContentFetcher into unified ExaClient class
- Add comprehensive docstrings for all methods and classes
- Rename fetch_and_parse() to fetch_content() for consistency
- Consolidate all Exa tests into test_exa_endpoint.py
- Add smart conditional testing (real API when key present, mocks when not)
- Remove deprecated files: exa_content_fetcher.py, test_exa_content_fetcher.py

Technical Benefits:
- Eliminate ~100 lines of duplicate code
- Single source of truth for Exa API integration
- Improved maintainability and consistent error handling
- Enhanced test coverage with real API validation

All privacy protections maintained. 20 tests passing.

robbyrob42 committed:
Remove user story file from repository

Keep user story documentation local only, not in version control.

robbyrob42 committed:
Fix Python import paths for typecheck compatibility

- Update imports in test_exa_endpoint.py to use relative paths (pro.context, pro.exa)
- Fix patch decorators to use correct module paths for mocking
- Update test_pro_tools.py to mock unified exa_client instead of deprecated exa_content_fetcher
- Resolves ty typecheck errors while maintaining all test functionality

robbyrob42 commented:
@cjroth all corrections implemented as requested. I grok the process and expectation now, and this should be easier for you to consume. I appreciate your feedback.

robbyrob42 committed:
Remove WebContentFetcher fallback and return Exa responses directly

- Remove fallback logic from fetch_content_endpoint per product requirements
- Return Exa error messages directly to users instead of falling back
- Clean up unused WebContentFetcher imports and initialization
- Update tests to use only mocking patches, remove API key dependent tests
- Fix error detection to properly handle Exa's string-based error responses
- All 84 backend tests passing, including comprehensive Exa endpoint coverage

This addresses the bug where Exa errors were incorrectly treated as successful
content, ensuring proper error handling without silent fallbacks.

cjroth commented on backend/pro/routes.py:
```diff
@@ -70,14 +69,29 @@ async def search_exa_endpoint(request: SearchRequest) -> SearchResponse:
-        """Fetch and parse content from a webpage URL"""
-        try:
-            ctx = SimpleContext()
-            content = await fetcher.fetch_and_parse(request.url, ctx)
```
Just realized - we need to remove the old `WebContentFetcher` code since it's no longer used as well

cjroth commented on backend/pro/exa.py:
```diff
@@ -106,8 +150,126 @@ async def search(
+            traceback.print_exc(file=sys.stderr)
+            return f"Error: An unexpected error occurred while fetching the webpage ({str(e)})"
+
+    def format_search_results_for_llm(self, results: list[dict[str, Any]]) -> str:
```
Do we really want to keep this? I thought exa was basically intended for use by LLMs and would return results in more or less a good format for LLMs as it is?

cjroth commented on backend/pro/exa.py:
```diff
@@ -106,8 +150,126 @@ async def search(
+        try:
+            headers = {"x-api-key": self.api_key, "Content-Type": "application/json"}
+
+            # Exa contents request payload - only the URL is sent to Exa
```
I think we should clean up all of these comments - I don't think they add much value but are adding visual clutter

robbyrob42 committed:
Remove unnecessary inline comments from exa.py per review feedback

robbyrob42 committed:
Refactor: Replace custom ExaClient with official exa-py SDK

- Added exa-py>=1.0.0 to dependencies
- Replaced custom ExaClient implementation with official SDK
- Removed format_search_results_for_llm method (SDK provides LLM-optimized format)
- Updated routes.py to use new SDK-based functions
- Simplified test suite to work with SDK implementation
- All tests passing (82/82)

This addresses PR review feedback about using Exa's native LLM formatting
and reduces code maintenance burden by using the official SDK.

robbyrob42 committed:
Fix linting issues for CI compliance

robbyrob42 committed:
Apply ruff formatting for CI compliance

robbyrob42 commented:
@cjroth moved to exa-labs/exa-py package. 
They: 
- format search results properly, 
- handle error handling and retry without effort from us, 
- enable unit tests without API key present so that we don't have to mock the api, and 
- have other tool classes and methods for future extension. 

sorry for the multiple reviews but this is clearly the correct ending location for us

cjroth commented on backend/pro/exa.py:
```diff
@@ -1,129 +1,117 @@
 from typing import Any
 
-import httpx
+from exa_py import Exa
```
OH, good call. Idk why we didn't just do that in the first place haha

cjroth replied:
Actually now that we're just using Exa's actual client - any reason we still need this exa class at all? I feel like we could just create and call the exa instance directly from the routes

robbyrob42 replied:
Yeah, this is me being too risk-averse to change while I was learning how the product works, but now that I am fully familiar, doing these things manually is clearly the most risky option. 

I'm open to removing the abstraction. I'll review this closely and see if there's any significant benefit to staying abstracted away from routes.py and either make the change or comeback for a chat

robbyrob42 replied:
the thin wrapper has a minimal cost overhead and, while not a ton of benefit right now, when we are adding integration environments or bringing in a new exa tool in the future, it will keep the route handling and functional operations separated. most other tools have a wrapper so that keeps the pattern consistent. suggest we keep for now.
